File: algo/quick_sort.py
def quick_sort(arr, s, e):
    if e - s + 1 <= 1:
        return arr
    
    pivot = arr[e]
    left = s

    for i in range(s, e):
        if arr[i] < pivot:
            arr[i], arr[left] = arr[left], arr[i]
            left += 1
    
    arr[e], arr[left] = arr[left], pivot

    quick_sort(arr, s, left - 1)
    quick_sort(arr, left + 1, e)

    return arr

def test_quick_sort():
    arr = [1, 3, 2, 4, 5]
    arr = quick_sort(arr, 0, len(arr) - 1)
    assert arr == [1, 2, 3, 4, 5]

    arr = [5, 4, 3, 2, 1]
    arr = quick_sort(arr, 0, len(arr) - 1)
    assert arr == [1, 2, 3, 4, 5]

    arr = [1, 2, 3, 4, 5]
    arr = quick_sort(arr, 0, len(arr) - 1)
    assert arr == [1, 2, 3, 4, 5]

    arr = [1]
    arr = quick_sort(arr, 0, len(arr) - 1)
    assert arr == [1]

    arr = []
    arr = quick_sort(arr, 0, len(arr) - 1)
    assert arr == []

    print("All tests pass")

test_quick_sort()

File: algo/quick_sort.js
function quickSort(arr, s, e) {
  if (e - s + 1 <= 1) return arr;

  let pivot = arr[e];
  let left = s;

  for (let i = s; i < e; i++) {
    if (arr[i] < pivot) {
      let tmp = arr[i];
      arr[i] = arr[left];
      arr[left] = tmp;
      left++;
    }
  }

  arr[e] = arr[left];
  arr[left] = pivot;

  quickSort(arr, s, left - 1);
  quickSort(arr, left + 1, e);

  return arr;
}

function test_quick_sort() {
  const assert = require('assert');
  const list = [1, 4, 2, 3, 5];
  let arr = quickSort(list, 0, list.length - 1);

  assert(arr[0] === 1);
  assert(arr[1] === 2);
  assert(arr[2] === 3);
  assert(arr[3] === 4);
  assert(arr[4] === 5);

  console.log('All test cases pass');
}

test_quick_sort();
